### PR TITLE
Re-org Lessons 7/8

### LINKED ISSUE #69: Update Lesson 7: reorganize into smaller units, interleave exercises w/ text
Currently the lesson is structured to have a lengthy section of informational text up-front, then a series of lessons at the end. Rather, interleave the instructional text and lessons, so that hands-on exercises happen immediately after covering the related information. 

Outline should be: 
- Title: **Lesson 7: Intro to Agents** 
- Goals: Explicitly describe Target Audience, Pre-Requisites, Intents.
- The Sensu Agent
- Exercise 1: Install the Sensu Agent
- Authentication and Communication
- Configuring the Agent
- Exercise 2: Customize Agent Configuration
- Discussion: Entity management, agent ingest API, deadman's switch, keepalives, other considerations 
- Learn More: Links to more reading 
- Next Steps: Workshop Navigation (standard links: feedback, next step, prev step)

---

NOTE: This is part of an effort to standardize the structure of all the lessons to look something like this (with the associated quantitative lexicographical limits):

Lesson Structure (500-1000 words):
- Task Based Titles/Structure: “Creating a new Handler” (5-7 words)
- Goals: Explicitly describe Target Audience, Pre-Requisites, Intents. (50-100 words)
- Exercises: One more more exercise (see _Exercise Structure_ below)
- Discussion: Move any edge cases, extra explanation, caveats, related features, advanced uses cases, etc to this section. (50-100 words)
- Learn More: Links to more reading (3-7 links)
- Next Steps: Workshop Navigation (standard links: feedback, next step, prev step)

Exercise Structure:
- Task: short description of the task (25-50 words)
- Solution: short description of how we’ll solve the problem/complete the task (25-50 words)
- Steps: Numbered steps 1-5, what to do. (50-100 words per step)

Sentences should be 5-7 words.
Paragraphs should be 2-5 sentences long (10-35 words).
Sections should be ~50-100 words long (10-20 sentences, or ~3-5 paragraphs).
Lessons should be 500-1000 words long (above structure is within those bounds)
Lessons should have 3-5 exercises
Exercises should have 3-5 steps
Exercises should be ~250 words long, using about 50 words per step.

Example: Lower end of range = 705 words.
- Title: 5 words
- Goals: 50 words
- Exercise 1: Task: 25 words
- Exercise 1: Solution: 25 words
- Exercise 1: Step 1: 50 words
- Exercise 1: Step 2: 50 words
- Exercise 1: Step 3: 50 words
- Exercise 2: Task: 25 words
- Exercise 2: Solution: 25 words
- Exercise 2: Step 1: 50 words
- Exercise 2: Step 2: 50 words
- Exercise 2: Step 3: 50 words
- Exercise 3: Task: 25 words
- Exercise 3: Solution: 25 words
- Exercise 3: Step 1: 50 words
- Exercise 3: Step 2: 50 words
- Exercise 3: Step 3: 50 words
- Discussion: 50 words
- Learn More: 3 links (n/a word count)
- Next Steps: feedback, previous, next (n/a word count)

Quantitative Metric Goals (as determined by this [Flesch-Kincaid calculator](https://goodcalculators.com/flesch-kincaid-calculator/)):
- Flesh-Kincaid Grade Level: 8-10
- Flesch Reading Ease Score: 40-60
- Average Words per Sentence: 4-7
- Average Syllables per Word: 1-2
- Sentences: 150-200
- Words: 750-1000

### PR DESCRIPTION
**Pull request by thoward**
Closes #69 and #70. 

This PR re-orgs Lessons 7/8 in the workshop. However, this is incomplete. 

Notably in Lesson 8, there's some TODOs that need to be addressed before merge, and there's some content about Entities in Lesson 7 that should be moved into a new lesson (or to an appendix if we don't want to renumber lessons/flow by adding a new lesson). Need to discuss plan there.

### CODE REVIEW

File: lessons/operator/07/README.md [hunk 1]
```diff
@@ -1,526 +1,155 @@
...
+For optimal network throughput, agents will attempt to negotiate the use of [Protobuf serialization](https://en.m.wikipedia.org/wiki/Protocol_Buffers) when communicating with a backend that supports it. Communication is via clear text by default.
 
-  [Learn more (reference docs)](https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-entities/entities/#manage-agent-entities-via-the-agent)
+### EXERCISE 1: Install the Sensu Agent
 
-**NOTE:** All agent entities are _created_ using the `sensu-agent` configuration; i.e. the _initial_ configuration for API-managed entities is provided by the `sensu-agent`, but subsequent edits made to the `sensu-agent` (via config flags, config file, or environment variables) will be ignored unless the `--agent-managed-entity` flag is set.
+#### Scenario
 
-## Advanced topics
+You have a server, container, or other device or piece of software that you want to manage with Sensu.
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: "You have a server, container, connected device, or service"

File: lessons/operator/07/README.md [hunk 2]
```diff
@@ -1,526 +1,155 @@
...
 
-### Proxy Entities
+#### Solution
 
-A "proxy entity" is any Sensu entity resource that is not actively under management by a Sensu Agent.
-Proxy entities have an `entity_class` of "proxy" and can be used to represent any resource under management by Sensu, though proxy entities are generally used to represent resources where we can't run an agent (e.g.  certain IoT/connected devices, or network devices), or resources that may emit observability data directly to Sensu without the need for an agent (e.g. applications or serverless functions).
+Install the the Sensu Agent on the system. 
+The agent runs as a separate process that observes your system.
```
Comment (calebhailey) [inline lessons/operator/07/README.md:53]: Strike this sentence.

File: lessons/operator/07/README.md [hunk 3]
```diff
@@ -1,526 +1,155 @@
...
+> Make sure you have the following environment variables set:
+>
+> - `SENSU_VERSION`
+> - `SENSU_BUILD`
+> - `SENSU_BACKEND_URL`
+> - `SENSU_NAMESPACE`
+> - `SENSU_USER`
+> - `SENSU_PASSWORD`
+> 
+> If any are missing, review the environment setup from [Lesson 3: Using the Sesnu CLI](../03/README.md#readme).
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: s/Sesnu/Sensu/

File: lessons/operator/07/README.md [hunk 4]
```diff
@@ -1,526 +1,155 @@
...
-A failed backup job will result in a TTL event without the need for any if/then/else conditional logic in the script (i.e. no need to also send an event if the script fails) – the mere absence of an "OK" event is all that is needed.
+      **Linux (amd64):**
 
-#### Examples
+      ```shell
+      curl -LO https://s3-us-west-2.amazonaws.com/sensu.io/sensu-go/${SENSU_VERSION}/sensu-go_${SENSU_VERSION}_linux_amd64.tar.gz 
+      ```
 
-**Bash**
+      **Linux (arm64):**
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: Let's take out ARM. Not necessary.

File: lessons/operator/07/README.md [hunk 5]
```diff
@@ -1,526 +1,155 @@
...
+      ```powershell
+      Invoke-WebRequest `
+       -Uri "https://s3-us-west-2.amazonaws.com/sensu.io/sensu-go/${Env:SENSU_VERSION}/sensu-go-agent_${Env:SENSU_VERSION}.${Env:SENSU_BUILD}_en-US.x64.msi" `
+       -OutFile "${Env:UserProfile}\sensu-go-agent_${Env:SENSU_VERSION}.${Env:SENSU_BUILD}_en-US.x64.msi"
+      ```
 
-Dead mans switches are useful for monitoring jobs like nighly backup jobs (e.g. a bash script scheduled via a cron job).
-A simple one-liner at the end of the backup script can be used to report on the backup status (e.g. `"Backup completed successfully. Backup data is available at <backup location URI>.`) with a ~25 hour TTL to account for long running backup jobs.
-A failed backup job will result in a TTL event without the need for any if/then/else conditional logic in the script (i.e. no need to also send an event if the script fails) – the mere absence of an "OK" event is all that is needed.
+      **Linux (amd64):**
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: s/Linux (amd64)/Linux/

File: lessons/operator/07/README.md [hunk 6]
```diff
@@ -1,526 +1,155 @@
...
 
-### Platform detection
+      ```powershell
+      msiexec.exe /i "${Env:UserProfile}\sensu-go-agent_${Env:SENSU_VERSION}.${Env:SENSU_BUILD}_en-US.x64.msi" /qr
+      rm "${Env:UserProfile}\sensu-go-agent_${Env:SENSU_VERSION}.${Env:SENSU_BUILD}_en-US.x64.msi"
+      ${Env:Path} += ";C:\Program Files\Sensu\sensu-agent\bin"
+      ```
 
-==TODO==
+      **Linux (amd64):**
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: s/Linux (amd64)/Linux/

File: lessons/operator/07/README.md [hunk 7]
```diff
@@ -1,526 +1,155 @@
...
 
-### Command allow list
+      ```shell
+      sudo -E tar -xzf sensu-go_${SENSU_VERSION}_linux_amd64.tar.gz -C /usr/bin/
+      rm sensu-go_${SENSU_VERSION}_linux_amd64.tar.gz
+      ```
 
-==TODO==
+      **Linux (arm64):**
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: Remove

File: lessons/operator/07/README.md [hunk 8]
```diff
@@ -557,36 +186,119 @@ The [official Sensu Go installation documentation](https://docs.sensu.io/sensu-g
    --password ${SENSU_PASSWORD}
    ```
 
-   Verify that your agent is running and connected to the Sensu Backend by consulting `sensuctl` (e.g. `sensuctl entity list`) or the Sensu web app.
+   Verify that your agent is running and connected to the backend by running `sensuctl entity list`.
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: Let's make this a more prominent step to complete this exercise. The user should expect to see an entity named "workshop" in the output.

File: lessons/operator/07/README.md [hunk 9]
```diff
@@ -557,36 +186,119 @@ The [official Sensu Go installation documentation](https://docs.sensu.io/sensu-g
...
+
+**NEXT:** If `sensu-agent` has successfully connected to your backend, you're ready to move on to the next exercise.
+
+## Configuring the Agent
+
+Some agent configuration parameters are used to manage the behavior of the agent, such as TLS certificates, authentication, and subscriptions.
+Others are used to configure aspects of the *entity* representation, such as labels and annotations.
```
Comment (calebhailey) [inline lessons/operator/07/README.md:194]: TODO: mention the lesson when we will review Entities in greater detail.

File: lessons/operator/07/README.md [hunk 10]
```diff
@@ -557,36 +186,119 @@ The [official Sensu Go installation documentation](https://docs.sensu.io/sensu-g
...
+If a configuration value is set in multiple places, it will be overrided with the following priority:
+
+1. **Command-line Option** (highest)
+2. **Environment Variable**
+3. **Configuration File**
+4. **Default Value** (lowest)
+
+### Configuration Names and Formats
+
+All of the agent configuration options have the same name regardless of the way they are set, differing only in format.
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: s/same name/consistent naming/ ??

File: lessons/operator/07/README.md [hunk 11]
```diff
@@ -557,36 +186,119 @@ The [official Sensu Go installation documentation](https://docs.sensu.io/sensu-g
...
+
+The agent can be configured via command-line options, YAML config file, or environment variables.
+If a configuration value is set in multiple places, it will be overrided with the following priority:
+
+1. **Command-line Option** (highest)
+2. **Environment Variable**
+3. **Configuration File**
+4. **Default Value** (lowest)
+
+### Configuration Names and Formats
```
Comment (calebhailey) [inline lessons/operator/07/README.md]: Move to "discussion"??

File: lessons/operator/08/README.md [hunk 12]
```diff
@@ -1,30 +1,32 @@
...
+  - [EXERCISE 3: Tagging and Handling Metrics Using Checks](#exercise-3-tagging-and-handling-metrics-using-checks)
+- [Discussion](#discussion)
+  - [Subscriptions, Loose Coupling, and Elastic Infrastructure](#subscriptions-loose-coupling-and-elastic-infrastructure)
+  - [Dead Man Switches (TTLs)](#dead-man-switches-ttls)
+  - [Proxy Checks (Pollers)](#proxy-checks-pollers)
+- [Learn More](#learn-more)
+- [Next Steps](#next-steps)
 
-## Overview
+## Goals
```
Comment (calebhailey) [inline lessons/operator/08/README.md:21]: TODO? LOL

File: lessons/operator/08/README.md [hunk 13]
```diff
@@ -54,174 +56,44 @@ spec:
...
 That's the entire specification (more or less)!
-Service checks have provided sustained value thanks to this incredibly simple specification, providing tremendous extensibility.
-In fact, service checks can be written in any programming language in the world (including simple Bash and MS DOS scripts).
+Service checks remain very useful because their simple specification makes it easy to extend monitoring to any area.
+Service checks can be written in any programming or scripting language, including Bash, PowerShell, and MS-DOS scripts.
+
+### Subscriptions and Check Scheduling
 
-## Scheduling
+In Sensu, *subscriptions* are equivalent to topics in a traditional [pub/sub](https://en.wikipedia.org/wiki/Publish–subscribe_pattern) model. Agents are the subscribers and the backend is the publisher.
```
Comment (calebhailey) [inline lessons/operator/08/README.md]: Line break after the first sentence? 

And can we reverse the second sentence here? "Backends publish, agents subscribe"?

Comment (calebhailey) [reply lessons/operator/08/README.md]: Screen Shot 2021-10-13 at 11 15 39 AM

File: lessons/operator/08/README.md [hunk 14]
```diff
@@ -248,42 +120,72 @@ spec:
...
+Tokens are references to entity attributes and metadata, wrapped in double curly braces (`{{  }}`).
+Default values can also be provided as a fallback for [unmatched tokens](https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-schedule/tokens/#unmatched-tokens).
+
+**Examples:**
+
+- **`{{ .name }}`:** replaced by the target entity name
+- **`{{ .labels.url }}`:** replaced by the target entity "url" label
+- **`{{ .labels.disk_warning | default "85%" }}`:** replaced by the target entity "disk_warning" label; if the label is not set then the default/fallback value of `85%` will be used
+
+==TODO: add an example check config w/ .labels.disk_warning label==
```
Comment (calebhailey) [inline lessons/operator/08/README.md]: Delete. This example isn't necessary thanks to the content re-org

File: lessons/operator/08/README.md [hunk 15]
```diff
@@ -313,48 +215,206 @@ Let's modify our check from the previous exercise using some Tokens.
...
 
-   Verify that the Check was successfully created using the `sensuctl check list` command:
+   Verify that the check was successfully created:
 
    ```shell
    sensuctl check info disk-usage --format yaml
    ```
 
-## EXERCISE 3: collecting metrics with Sensu Checks
+1. TODO: Add annotation to an entity?
```
Comment (calebhailey) [inline lessons/operator/08/README.md]: TODO - update the agent configuration to add new annotations

File: lessons/operator/08/README.md [hunk 16]
```diff
@@ -313,48 +215,206 @@ Let's modify our check from the previous exercise using some Tokens.
...
+
+## Metrics Collection and Extractions
```
Comment (calebhailey) [inline lessons/operator/08/README.md]: s/Extractions/Processing/

**Review by calebhailey — COMMENTED** [review]
WIP Review Comments

File: lessons/operator/08/README.md [hunk 17]
```diff
@@ -313,48 +215,206 @@ Let's modify our check from the previous exercise using some Tokens.
...
+Metric tag values can be provided as strings, or [tokens](https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-schedule/tokens/) which can be used for generating dynamic tag values.
+
+### Output Metric Handlers
+
+In addition to `output_metric_format`, Sensu checks also provide configuration for dedicated `output_metric_handlers`, which are event handlers that are specially optimized for processing metrics.
+If an event containing metrics is configured with one or more `output_metric_handlers`, a copy of the event is forwarded to the metric handler prior to Sensu's own event persistence; this specialized handling is implemented as a performance optimization to prioritize metric processing.
+
+> **NOTE:** Checks may be configured with multiple `handlers` and `output_metric_handlers`, enabling service health checking, alerting, _and_ metrics collection in a single check.
+
+### EXERCISE 3: Tagging and Handling Metrics Using Checks
```
Comment (calebhailey) [inline lessons/operator/08/README.md]: s/Tagging and Handling/Processing/

File: lessons/operator/08/README.md [hunk 18]
```diff
@@ -313,48 +215,206 @@ Let's modify our check from the previous exercise using some Tokens.
...
+
+#### Scenario
+
+You have some existing metrics providers which provide output in Nagios format, which you want to store in InfluxDB. You also want to capture some additional information about the entity and have the metrics, along with it's context be handled by the pipeline.
```
Comment (calebhailey) [inline lessons/operator/08/README.md]: "existing metrics providers" might be a little too abstract here; perhaps "existing monitoring plugins" ?

Comment (calebhailey) [inline lessons/operator/08/README.md]: s/InfluxDB/a data platform like Sumo Logic or InfluxDB/

Comment (calebhailey) [inline lessons/operator/08/README.md]: Line break after the first sentence.

Comment (calebhailey) [inline lessons/operator/08/README.md]: Rephrase the second sentence to say "You also want to enrich collected metrics with information from the metric source (e.g. server or service metadata)." ??

File: lessons/operator/08/README.md [hunk 19]
```diff
@@ -313,48 +215,206 @@ Let's modify our check from the previous exercise using some Tokens.
...
+
+### EXERCISE 3: Tagging and Handling Metrics Using Checks
+
+#### Scenario
+
+You have some existing metrics providers which provide output in Nagios format, which you want to store in InfluxDB. You also want to capture some additional information about the entity and have the metrics, along with it's context be handled by the pipeline.  
+
+#### Solution
+
+This can be accomplished by configuring the check to expect Nagios format via the `output_metric_format` option, and configuring a metrics-specific storage handler via `output_metric_handlers`. We can also add additional metadata to the event using `output_metric_tags`.
```
Comment (calebhailey) [inline lessons/operator/08/README.md]: "to expect Nagios formatted metrics" ?

Comment (calebhailey) [reply lessons/operator/08/README.md]: and should we s/output_metric_format/output_metric_format: nagios_perfdata/ ??

File: lessons/operator/07/README.md (hunk 9 above)
Comment (thoward) [reply lessons/operator/07/README.md:194]: Need to figure out where that goes in the flow first, then can come back and mention it here.

**Review by calebhailey — APPROVED** [review]
LGTM!

**Comment by thoward** [comment]
Closes #70

**Pull request merged**